src/Futhark/Internalise/NormaliseSize.hs

```diff
@@ -502,57 +483,59 @@ arrowCleanType paramed (Scalar ty) =
+  | Just vn <- M.lookup (SizeExpr e) mapping =
+      Var (qualName vn) (Info $ typeOf e) (srclocOf e)
+expReplace mapping e = runIdentity $ astMap mapper e
   where
-    mapper = identityMapper {mapOnExp = expReplace mapping}
+    mapper = identityMapper {mapOnExp = pure . expReplace mapping}
 
 -- Normalise a ValBind
-removeExpFromValBind :: ValBind -> NormaliseM ValBind
-removeExpFromValBind valbind = do
-  scope <-
-    asks $
-      S.union (S.fromList $ mapMaybe unParamDim $ valBindTypeParams valbind)
-  (params', expNaming) <- runInnerNormaliser (mapM onPat $ valBindParams valbind) scope mempty
+normaliseValBind :: MonadFreshNames m => S.Set VName -> ValBind -> m (S.Set VName, ValBind)
+normaliseValBind prev_scope valbind = do
+  let scope =
+        prev_scope
+          <> S.fromList (mapMaybe unParamDim $ valBindTypeParams valbind)
+  (params', exp_naming) <-
+    runNormaliseM (mapM onPat $ valBindParams valbind) scope mempty
 
   let args = foldMap patNames params'
-      argsParams = M.elems expNaming
-  (rety', extNaming) <- runInnerNormaliser (hardOnRetType $ unInfo $ valBindRetType valbind) (scope <> args) expNaming
-  (rety'', (funArg, _)) <-
-    runWriterT (runReaderT (arrowArgRetType args rety') (scope, mempty))
-  let newParams = funArg `S.union` S.fromList argsParams
-      rety''' = arrowCleanRetType newParams rety''
+      args_params = M.elems exp_naming
+  (rety', extNaming) <-
+    runNormaliseM (hardOnRetType $ unInfo $ valBindRetType valbind) (scope <> args) exp_naming
+  let (rety'', (funArg, _)) =
+        runWriter (runReaderT (arrowArgRetType args rety') (scope, mempty))
+      new_params = funArg `S.union` S.fromList args_params
+      rety''' = arrowCleanRetType new_params rety''
       typeParams' =
         valBindTypeParams valbind
-          <> map (`TypeParamDim` mempty) (S.toList newParams)
-      expNaming' = M.filter (`S.member` newParams) extNaming
-
-  let scope' = scope `S.union` args `S.union` newParams
-  (body', expNaming'') <- runInnerNormaliser (expFree $ valBindBody valbind) scope' expNaming'
-  let newNames = expNaming'' `M.difference` expNaming'
-  body'' <- foldrM insertDimCalculus body' $ M.toList $ canCalculate scope' newNames
-  body''' <- expReplace expNaming' body''
-
-  pure $
-    valbind
-      { valBindRetType = Info rety''',
-        valBindTypeParams = typeParams',
-        valBindParams = params',
-        valBindBody = body'''
-      }
+          <> map (`TypeParamDim` mempty) (S.toList new_params)
+      exp_naming' = M.filter (`S.member` new_params) extNaming
+
+  let scope' = scope `S.union` args `S.union` new_params
+  (body', exp_naming'') <-
+    runNormaliseM (expFree $ valBindBody valbind) scope' exp_naming'
+  let new_names = exp_naming'' `M.difference` exp_naming'
+  body'' <-
+    expReplace exp_naming'
+      <$> foldrM insertDimCalculus body' (M.toList $ canCalculate scope' new_names)
+
+  pure
+    ( S.insert (valBindName valbind) scope,
```

Comment:
shouldn't it be prev_scope ? scope also has the size parameters of the current binding

Reply:
Yes.  Fixed.